### PR TITLE
Support latest version auto incremental

### PR DESCRIPTION
**Pull request by SophieGuo410**
This pr support:
1. for put request, if user provide LATEST for timestamp/monotonic schema dataset, we will issue a put with currentLatestVersion+1.
2. for put request, if user provide MAJOR/MINOR/PATCH for semantic schema dataset, we will issue a put with major/minor/patch version +1 depend on different option.
3. for get request, when user provide LATEST, will return the currentLatest dataset version.

### CODE REVIEW

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java [hunk 1]
```diff
@@ -397,6 +404,91 @@ public synchronized DatasetVersionRecord addDatasetVersions(int accountId, int c
...
+        //always retry with the latest version + 1 until it has been successfully uploading without conflict or failed with exception.
+        return addDatasetVersionHelper(accountId, containerId, dataset, version, timeToLiveInSeconds, creationTimeInMs,
+            datasetVersionTtlEnabled, versionNumber);
+      } catch (SQLException | AccountServiceException ex) {
+        // if the version already exist in database, retry with the new latest version +1.
+        if (!(ex instanceof SQLIntegrityConstraintViolationException)) {
+          throw ex;
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (JingQianCloud) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java:463]: Good point.

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java [hunk 2]
```diff
@@ -845,6 +958,87 @@ private long executeListVersionStatement(PreparedStatement statement, int accoun
...
+          patchVersion = 0;
+          if (majorVersion > MAX_SEMANTIC_MAJOR_MINOR_PATH_VERSION_VALUE) {
+            throw new IllegalArgumentException("The largest version for Semantic Major version should be less than "
+                + MAX_SEMANTIC_MAJOR_MINOR_PATH_VERSION_VALUE + ", currentMajorVersion: " + majorVersion);
+          }
+        } else if (MINOR.equals(version)) {
+          minorVersion += 1;
+          patchVersion = 0;
+          if (minorVersion > MAX_SEMANTIC_MAJOR_MINOR_PATH_VERSION_VALUE) {
+            throw new IllegalArgumentException("The largest version for Semantic Major version should be less than "
```
Comment (JingQianCloud) [inline ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java:1010]: nit: Major-> Minor

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java [hunk 3]
```diff
@@ -845,6 +958,87 @@ private long executeListVersionStatement(PreparedStatement statement, int accoun
...
+                + MAX_SEMANTIC_MAJOR_MINOR_PATH_VERSION_VALUE + ", currentMinorVersion: " + minorVersion);
+          }
+        } else {
+          patchVersion += 1;
+          if (patchVersion > MAX_SEMANTIC_MAJOR_MINOR_PATH_VERSION_VALUE) {
+            throw new IllegalArgumentException("The largest version for Semantic Major version should be less than "
```
Comment (JingQianCloud) [inline ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java:1016]: nit: Major->Patch

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java [hunk 4]
```diff
@@ -49,6 +49,12 @@ public class AccountDao {
   private final MySqlDataAccessor dataAccessor;
   private final MySqlAccountServiceConfig mySqlAccountServiceConfig;
   private final ObjectMapper objectMapper = new ObjectMapper();
+  private static final long MAX_TIMESTAMP_MONOTONIC_VERSION_VALUE = 9999999999L;
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java]: Updated.

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java:467]: Nice catch! Updated.

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java [hunk 5]
```diff
@@ -418,8 +510,15 @@ public synchronized DatasetVersionRecord getDatasetVersions(short accountId, sho
...
+      long versionValue;
+      if (isAutoCreatedVersionForDownload(version)) {
+        PreparedStatement listLatestVersionStatement = dataAccessor.getPreparedStatement(listLatestVersionSql, false);
+        versionValue =
+            executeListLatestVersionStatement(listLatestVersionStatement, accountId, containerId, datasetName);
+        version = convertVersionValueToVersion(versionValue, versionSchema);
+      } else {
+        versionValue = getVersionBasedOnSchema(version, versionSchema);
+      }
       DatasetVersionRecord result =
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java]: Sure. optimized the logic.

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java (hunk 2 above)
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java]: Fixed.

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java (hunk 3 above)
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java]: Fixed.

**Review by justinlin-linkedin — APPROVED** [review]

**Review by JingQianCloud — APPROVED** [review]

**Pull request merged**